### ISSUE TITLE
Turn off eager expression reordering

### ISSUE THREAD

**Issue by kwmsmith** [blaze/blaze#1542]
```python
In [2]: t = bz.symbol('t', 'var * {name: string, id: int}')

In [6]: t[['name']][t.name == 'Ying']
Out[6]: t[t.name == 'Ying'][['name']]

In [7]: t[['name', 'id']][['name']]
Out[7]: t[['name']]

In [8]: t[['name', 'id']][['id']]
Out[8]: t[['id']]

In [9]: t[['name', 'id']][t.id < 100][['id']]
Out[9]: t[t.id < 100][['id']]
```

We have usecases that build up expressions incrementally.  The final composite expression needs to be able to indicate which smaller expressions it is derived from.  Blaze's eager expression rewriting makes it impossible to do this at times.

One solution would be to turn off any eager expression rewriting or reordering, and instead do rewriting / reordering as a `compute()` time optimization.

**Comment by skrah** [comment]
@kwmsmith If eager rewriting is turned off, are expressions that are not commutative allowed to fail?

Example:

```psql
blaze_test=# select * from alice_bob;
 id | name  | amount 
----+-------+--------
  1 | Alice |    100
  2 | Bob   |    200
  3 | Alice |     50
(3 rows)

blaze_test=# select * from (select name from alice_bob) as t where amount < 100;
ERROR:  column "amount" does not exist
LINE 1: ...ct * from (select name from alice_bob) as t where amount < 1...
                                                             ^
blaze_test=# select name from (select * from alice_bob where amount < 100) as t;
 name  
-------
 Alice
(1 row)
```

Eager rewriting currently lets some computations succeed that would otherwise fail naturally like above.  The test suite contains some examples (e.g. test_spark).

**Comment by kwmsmith** [comment]
For this specific example, what is the blaze expression that's being rewritten / reordered here?

Is it something like:

```
t[t.amount < 100].name
```

When I repr this expression, it isn't being reordered -- it's symbolic representation looks identical to the expr as input.

The expression reordering I'm referring to here is entirely at the symbolic Blaze level *before* any `compute` happens.  It's fine if Blaze reorders an expression as an optimization pass, but that has to happen *during* `compute`, not when the original symbolic expr is constructed.

So, for instance, when I do a projection followed by a selection:
```
t[['name']][t.amount < 100]
```

The resulting symbolic expression is constructed as a selection, *then* a projection:

```
t[t.amount < 100][['name']]
```

*That* logic needs to be factored out and moved into `optimize`, and wherever else this is occurring.

For this specific example, the offending line is here:  https://github.com/blaze/blaze/blob/master/blaze/expr/expressions.py#L544

That logic needs to be moved into the `optimize` pass.

**Comment by kwmsmith** [comment]
One way to find where this is likely happening is wherever `_subs()` is called inside expression constructors.  I suspect and hope this is somewhat infrequent.

**Comment by skrah** [comment]
Yes, your example is one the lines that I removed:

    https://github.com/blaze/blaze/blob/master/blaze/expr/expressions.py#L544

But https://github.com/blaze/blaze/blob/master/blaze/compute/tests/test_spark.py#L241 seems to rely on this eager reordering.


So my question was: Given that this double-select spark example would also naturally fail in sql, can users be expected to write this instead?

`t[t['amount'] < 100]['name']`


Or is the commutativity part of the published API?

**Comment by kwmsmith** [comment]
> Yes, your example is one the lines that I removed

Did you remove it without moving that logic to `optimize`?  That logic / reordering does need to happen before the actual computation happens, otherwise we'll get the errors we're seeing.

One suggestion would be to make a new `reorder_projection()` optimization pass, and put it in `blaze/expr/optimize.py` -- you can follow the pattern for `lean_projection()`.  You'll have to ensure that this optimization is run during `compute`.

**Comment by skrah** [comment]
I didn't move the logic to `optimize`.  A related question is:

Currently the test suite seems to pass (except for two hard-coded sql output comparisons) when `optimize` is disabled:

```
    # 3. Re-optimize data and expressions
    optimize_ = False #  kwargs.get('optimize', optimize)
```

This issue would move essential reorderings into `optimize`. Should there perhaps be a separate pass for the essential reorderings?

**Comment by kwmsmith** [comment]
> This issue would move essential reorderings into optimize. Should there perhaps be a separate pass for the essential reorderings?

Yes, good point!  This is a good case for a new `transform()` function that takes an expression and returns the transformed expr, to be done right at the top of `compute()`.  It would need to happen before optimization, and we could put other expression transformations there as more accumulate.

**Comment by llllllllll** [comment]
There is precedent for making required expr rewrites in `optimize`. Instead of adding a new pass, we could rename optimize to `rewrite_expression` or something more clear.

**Comment by kwmsmith** [comment]
> There is precedent for making required expr rewrites in optimize. Instead of adding a new pass, we could rename optimize to rewrite_expression or something more clear.

OK, I'm +1 on the rename -- what about `transform_expr()`?  I'm used to multi-pass compilers that have a pipeline of transformations on an AST -- this strikes me as something very similar, and `transform_expr()`'s job would be to pass an input expression through the pipeline.

**Comment by llllllllll** [comment]
The multipass compiler is also what I am thinking of, so `transform` sounds good to me.

**Comment by kwmsmith** [comment]
+1 `transform`
